### PR TITLE
Reintroduce the fix for the reverted PR: Postblit has priority over copy constructor

### PR DESCRIPTION
**Pull request by RazvanN7**
This reintroduces the changes reverted in #11947 , however, it does no longer silently change code behavior.

Spec PR: https://github.com/dlang/dlang.org/pull/2931

cc @Geod24

### CODE REVIEW

**Comment by RazvanN7** [comment]
Autotester fail looks unrelated.

File: changelog/copy-constructor-priority.dd [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Copy constructor priority with regards to postblit
```
Comment (Geod24) [inline changelog/copy-constructor-priority.dd:1]:
```suggestion
Having both a copy constructor and a postblit is not deprecated
```

File: changelog/copy-constructor-priority.dd [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+Up until this release, the postblit had priority over the copy constructor,
+no matter if the former was user-defined or generated. This prevented new
+code that uses the copy constructor to interact with legacy library code
```
Comment (Geod24) [inline changelog/copy-constructor-priority.dd:5]:
```suggestion
code that uses the copy constructor to interact with library code
```

To be fair, copy ctors are still very experimental and not that used :)

File: changelog/copy-constructor-priority.dd [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// new code using copy constructor
+struct B
+{
+    A a;
+    this(const scope ref B) {}
+}
+---
+
+Will issue a deprecation warning at the definition of `B` stating that the postblit
+is going to be preferred to the copy constructor. To instruct the compiler that
```
Comment (Geod24) [inline changelog/copy-constructor-priority.dd:25]:
```suggestion
is going to be preferred to the copy constructor in a future release. To instruct the compiler that
```

File: changelog/copy-constructor-priority.dd [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+the copy constructor has to be used instead of the postblit, the developer has
+to disable the postblit for `struct B`.
+
+If `B` has both a user-defined postblit and a copy constructor (generated or user-defined),
+the postblit will continue to have priority.
```
Comment (Geod24) [inline changelog/copy-constructor-priority.dd:30]: I'd rephrase this a bit, saying that the user has two courses of actions to solve this deprecation warning:
- Explicitly add `@disable this(this);`, which instructs the compiler that the user wants the copy ctor to be used over the postblit;
- Remove or disable the copy constructor;
- Explicitly define a postblit;

And hopefully emphasis on the fact that the first solution should be preferred (it's likely to be the right one for most people).

File: src/dmd/dsymbolsem.d [hunk 5]
```diff
@@ -86,6 +86,8 @@ private FuncDeclaration buildPostBlit(StructDeclaration sd, Scope* sc)
     if (sd.isUnionDeclaration())
         return null;
 
+    bool hasUserDefinedPosblit = sd.postblits.dim && !sd.postblits[0].isDisabled ? true : false;
```
Comment (Geod24) [inline src/dmd/dsymbolsem.d:89]: `const`

File: src/dmd/dsymbolsem.d [hunk 6]
```diff
@@ -333,6 +335,27 @@ private FuncDeclaration buildPostBlit(StructDeclaration sd, Scope* sc)
...
+        // we have user defined postblit, so we prioritize it
+        if (hasUserDefinedPosblit)
+        {
+            sd.hasCopyCtor = false;
+            return xpostblit;
+        }
+        // we have fields with postblits, so print deprecations
+        if (xpostblit && !xpostblit.isDisabled())
+        {
+            deprecation(sd.loc, "`struct %s` has both a user defined copy constructor and fields with postblits.", sd.toChars);
```
Comment (Geod24) [inline src/dmd/dsymbolsem.d:350]:
```suggestion
            deprecation(sd.loc, "`struct %s` implicitly-generated postblit hides copy constructor", sd.toChars);
```

I don't know how to best express this. The original message is fine once you read the changelog entry, but without knowing the context, it's not obvious why it's a problem. Using "hide" (or whatever you feel is suitable) brings the idea that the copy constructor *should have priority*, as we want to eventually get rid of postblit (if that's still the plan ?).

File: src/dmd/dsymbolsem.d [hunk 7]
```diff
@@ -333,6 +335,27 @@ private FuncDeclaration buildPostBlit(StructDeclaration sd, Scope* sc)
...
+            deprecationSupplemental(sd.loc, "The field postblit will have priority over the copy constructor.");
+            deprecationSupplemental(sd.loc, "To change this, the postblit should be disabled for `struct %s`", sd.toChars());
```
Comment (Geod24) [inline src/dmd/dsymbolsem.d:352]:
```suggestion
            deprecationSupplemental(sd.loc, "Use `@disable this(this);`, define postblit explicitly, or remove or delete the copy constructor");
```

Perhaps also point to which fields trigger the postblit generation, if that's available ?

File: src/dmd/dsymbolsem.d [hunk 8]
```diff
@@ -466,16 +485,19 @@ private bool buildCopyCtor(StructDeclaration sd, Scope* sc)
...
+        {
+            .error(sd.loc, "`struct %s` may not define both a rvalue constructor and a copy constructor", sd.toChars());
+            errorSupplemental(rvalueCtor.loc,"rvalue constructor defined here");
+            errorSupplemental(cpCtor.loc, "copy constructor defined here");
+            return true;
+        }
+        else
+        {
+            return true;
+        }
```
Comment (Geod24) [inline src/dmd/dsymbolsem.d:500]: That `else` (not the return inside) is useless since the other branch returns

**Review by Geod24 — COMMENTED** [review]
Very nice!

File: changelog/copy-constructor-priority.dd (hunk 3 above)
Comment (RazvanN7) [reply changelog/copy-constructor-priority.dd:25]: But it's during this release, no?

File: src/dmd/dsymbolsem.d (hunk 7 above)
Comment (RazvanN7) [reply src/dmd/dsymbolsem.d:352]: I thought about that, but since the fix is to disable the postblit I think it is not relevant to output the fields.

File: changelog/copy-constructor-priority.dd (hunk 4 above)
Comment (RazvanN7) [reply changelog/copy-constructor-priority.dd:30]: The copy constructor can be overloaded so I think that for the ease of implementation we can still trigger the deprecation even if all copy constructors are disabled. You can simply add `this(this) {}` to achieve the same effect as disabling all copy constructors.

**Comment by RazvanN7** [comment]
@Geod24 I addressed your comments.

**Review by Geod24 — APPROVED** [review]
